File: src/graph.c
#include "./includes/main.h"

int **iniGraph(void) {
	int **graph = calloc(MAPSIZE,sizeof(int*));
	assert(graph);
	for (int i=0; i < MAPSIZE; i++)
		graph[i] = calloc(MAPSIZE,sizeof(int));
	return graph;
}

void delGraph(int **graph) {
	for (int i=0 ; i < MAPSIZE ; i++)
		free(graph[i]);
	free(graph);
}

void prtGraph(int **graph) {
	bool line;
	for (int i=0; i < MAPSIZE; i++) {
		line = false;
		for (int j=0; j < MAPSIZE; j++) {
			if (graph[i][j]) {
				if (!line) {
					printf(" %d ->", i);
					line = true;
				}
				printf(" %d", j);
			}
		}
		if (line) printf("\n");
	}
}

void gGraph(State *hashmap, int **graph) {
	int parent;
	for (int i=0; i < MAPSIZE; i++) {
		if (hashmap[i].mapped && hashmap[i].src != -1) {
			parent = hashmap[i].src;
			graph[parent][i] = 1;
		}
	}
}

void genDot(State *hashmap, int **graph, char *file) {
	char output[0x100];
	strcpy(output,"./resources/graphs/");
	strcat(output,file);
	strcat(output,".dot");

	FILE *fp = fopen(output,"w");
	assert(fp);

	int m,c;

	fprintf(fp,"digraph {\n");
	for (int i=0; i < MAPSIZE; i++) {

		if (hashmap[i].mapped) {

			if (!hashmap[i].b) {
				m = M - hashmap[i].m;
				c = C - hashmap[i].c;
			}
			else {
				m = hashmap[i].m;
				c = hashmap[i].c;
			}

			fprintf(fp,"\t\"%d\"[label=",i);
			fprintf(fp,"\"(%d,%d,%d)\"",m,c,hashmap[i].b);

			if (hashmap[i].dinner)
				fprintf(fp,",style=filled,fillcolor=red]\n");
			else
				fprintf(fp,"]\n");
		}
	}

	for (int i=0; i < MAPSIZE; i++) {
		for (int j=0; j < MAPSIZE; j++) {

			if (graph[i][j]) {

				if (!hashmap[i].b) {
					m = hashmap[j].m - hashmap[i].m;
					c = hashmap[j].c - hashmap[i].c;
				}
				else {
					m = hashmap[i].m - hashmap[j].m;
					c = hashmap[i].c - hashmap[j].c;
				}

				if (i != j) {
					fprintf(fp,"\t\"%d\" -> \"%d\"",i,j);
					fprintf(fp,"[label=\"%d,%d\",",m,c);
					fprintf(fp,"weight=\"%d,%d\"]\n",m,c);
				}
			}
		}
	}

	fprintf(fp,"}");
	fclose(fp);
}

void genViz(State *hashmap, int **graph, char *file) {
	char output[0x100];
	strcpy(output,"./resources/graphs/");
	strcat(output,file);
	strcat(output,".js");

	FILE *fp = fopen(output,"w");
	assert(fp);

	int m,c;

	fprintf(fp,"const gData = {\n");
	fprintf(fp,"\tnodes: [\n");
	for (int i=MAPSIZE-1; i >= 0; i--) {

		if (hashmap[i].mapped) {

			if (!hashmap[i].b) {
				m = M - hashmap[i].m;
				c = C - hashmap[i].c;
			}
			else {
				m = hashmap[i].m;
				c = hashmap[i].c;
			}

			fprintf(fp,"\t{ id: %d, name: \"(",i);
			fprintf(fp,"%d,%d,%d)\"",m,c,hashmap[i].b);
			//fprintf(fp,", val: %d", hashmap[i].deep);

			if (hashmap[i].m == 0 && hashmap[i].c == 0)
				fprintf(fp,", color: \"green\" },\n");
			else if (hashmap[i].m == M && hashmap[i].c == C)
				fprintf(fp,", color: \"green\" },\n");
			else if (hashmap[i].dinner)
				fprintf(fp,", color: \"red\" },\n");
			else
				fprintf(fp," },\n");
		}
	}

	fprintf(fp,"\t],\n\tlinks: [\n");

	for (int i=0; i < MAPSIZE; i++) {
		for (int j=0; j < MAPSIZE; j++) {

			if (graph[i][j]) {

				if (!hashmap[i].b) {
					m = hashmap[j].m - hashmap[i].m;
					c = hashmap[j].c - hashmap[i].c;
				}
				else {
					m = hashmap[i].m - hashmap[j].m;
					c = hashmap[i].c - hashmap[j].c;
				}

				if (i != j) {
					fprintf(fp,"\t{ target: %d, source: %d,",j,i);
					fprintf(fp," name: \"%d,%d\" },\n",m,c);
				}
			}
		}
	}
	
	fprintf(fp,"\t]\n};");
	fclose(fp);
}

File: teste/resources/graph.js
const gData = {
    nodes: [
	{ id: 0, name: "(3,3,0)" },
	{ id: 1, name: "(3,2,0)" },
	{ id: 2, name: "(3,1,0)" },
	{ id: 10, name: "(2,3,0)", color: "red" },
	{ id: 11, name: "(2,2,0)" },
	{ id: 12, name: "(2,1,0)", color: "red" },
	{ id: 13, group: 133, name: "(2,0,0)", color: "red" },
	{ id: 20, name: "(1,3,0)", color: "red" },
	{ id: 21, name: "(1,2,0)", color: "red" },
	{ id: 22, group: 133, name: "(1,1,0)" },
	{ id: 23, group: 133, name: "(1,0,0)", color: "red" },
	{ id: 30, name: "(0,3,0)" },
	{ id: 31, group: 133, name: "(0,2,0)" },
	{ id: 32, group: 133, name: "(0,1,0)" },
	{ id: 102, name: "(0,2,1)" },
	{ id: 103, name: "(0,3,1)" },
	{ id: 111, name: "(1,1,1)" },
	{ id: 112, name: "(1,2,1)", color: "red" },
	{ id: 113, name: "(1,3,1)", color: "red" },
	{ id: 121, name: "(2,1,1)", color: "red" },
	{ id: 122, name: "(2,2,1)" },
	{ id: 123, name: "(2,3,1)", color: "red" },
	{ id: 131, name: "(3,1,1)" },
	{ id: 132, name: "(3,2,1)" },
	{ id: 133, name: "(3,3,1)" }
	],
    links: [
	 { target: 1, source: 102 },
	 { target: 1, source: 111 },
	 { target: 2, source: 103 },
	 { target: 11, source: 112 },
	 { target: 11, source: 113 },
	 { target: 11, source: 121 },
	 { target: 11, source: 122 },
	 { target: 22, source: 123 },
	 { target: 22, source: 132 },
	 { target: 30, source: 131 },
	 { target: 102, source: 0 },
	 { target: 103, source: 1 },
	 { target: 111, source: 10 },
	 { target: 122, source: 2 },
	 { target: 131, source: 11 },
	 { target: 131, source: 20 },
	 { target: 132, source: 12 },
	 { target: 132, source: 21 },
	 { target: 132, source: 30 },
	 { target: 133, source: 13 },
	 { target: 133, source: 22 },
	 { target: 133, source: 23 },
	 { target: 133, source: 31 },
	 { target: 133, source: 32 }
	]
};

const Graph = ForceGraph()
	(document.getElementById('graph'))
  	.graphData(gData);